Dockerfile

```diff
@@ -0,0 +1,48 @@
+# Copyright 2022 The KCP Authors.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+# Build the binary
+FROM golang:1.18 AS builder
+
+WORKDIR /build-dir
+# Copy the go source
+COPY . .
+# cache deps before building and copying source so that we don't need to re-download as much
+# and so that source changes don't invalidate our downloaded layer
```

Comment:
https://github.com/kcp-dev/kcp/blob/main/Dockerfile has this comment. BTW I do not understand why it can lead to "we don't need to re-download", do you have idea?
Another issue is, https://github.com/kcp-dev/kcp/blob/main/Dockerfile really does it **before** its line "# Copy the go source", but here we does it **after** the line "COPY . .", so the comment "cache deps **before** ... copying source" is a little not the actuality. It is better to make comment consist to the actuality.

Reply:
Good question! I think it only works when we build it on our local machine when the go mod depedency not changes(most time). The comments need to update, I'll update it. Thanks!

Reply:
Except this, the remaining parts LGTM

Reply:
> I think it only works when we build it on our local machine when the go mod depedency not changes(most time)
Yes this is my understanding, too.  Other build systems (like github actions) can be configured to cache container layers to speed up builds, but without that caching configuration, every layer will get recreated for every build.